src/util/amount.rs

```diff
@@ -0,0 +1,876 @@
+        assert_eq!(p("5 satoshi"), Ok(Amount::from_sat(5)));
+        assert_eq!(p("0.10000000 BTC"), Ok(Amount::from_sat(100_000_00)));
+        assert_eq!(p("100 bits"), Ok(Amount::from_sat(10_000)));
+    }
+
+    #[test]
+    fn to_string_with_denomination_from_str_roundtrip() {
+        use super::Denomination as D;
+
+        let amt = Amount::from_sat(42);
+        assert_eq!(Amount::from_str(&amt.to_string_with_denomination(D::Bitcoin)), Ok(amt));
+        assert_eq!(Amount::from_str(&amt.to_string_with_denomination(D::MilliBitcoin)), Ok(amt));
+        assert_eq!(Amount::from_str(&amt.to_string_with_denomination(D::MicroBitcoin)), Ok(amt));
+        assert_eq!(Amount::from_str(&amt.to_string_with_denomination(D::Bit)), Ok(amt));
+        assert_eq!(Amount::from_str(&amt.to_string_with_denomination(D::Satoshi)), Ok(amt));
+        assert_eq!(Amount::from_str(&amt.to_string_with_denomination(D::MilliSatoshi)), Ok(amt));
+    }
+
+    #[cfg(feature = "serde")]
+    #[test]
+    fn serde_as_sat() {
+        use serde::{Deserialize, Serialize};
+
+        #[derive(Serialize, Deserialize, PartialEq, Debug)]
+        struct T {
+            #[serde(with = "::util::amount::serde::as_sat")]
+            pub amt: Amount,
+        }
+
+        serde_test::assert_tokens(
+            &T {
+                amt: Amount::from_sat(123456789),
+            },
+            &[
+                serde_test::Token::Struct {
+                    name: "T",
+                    len: 1,
+                },
+                serde_test::Token::Str("amt"),
+                serde_test::Token::U64(123456789),
+                serde_test::Token::StructEnd,
+            ],
+        );
+    }
+
+    #[cfg(feature = "serde")]
+    #[test]
+    fn serde_as_btc() {
+        use serde::{Deserialize, Serialize};
+        use serde_json;
+
+        #[derive(Serialize, Deserialize, PartialEq, Debug)]
+        struct T {
+            #[serde(with = "::util::amount::serde::as_btc")]
+            pub amt: Amount,
+        }
+
+        let orig = T {
+            amt: Amount::from_sat(21_000_000__000_000_01),
+        };
+
+        let t: T = serde_json::from_str("{\"amt\": 21000000.00000001}").unwrap();
+        assert_eq!(t, orig);
+
+        let value: serde_json::Value =
+            serde_json::from_str("{\"amt\": 21000000.00000001}").unwrap();
+        assert_eq!(t, serde_json::from_value(value).unwrap());
+
+        // errors
+        let t: Result<T, serde_json::Error> = serde_json::from_str("{\"amt\": -42.0}");
+        assert!(t.unwrap_err().to_string().contains(&ParseAmountError::Negative.to_string()));
+        let t: Result<T, serde_json::Error> = serde_json::from_str("{\"amt\": 1000000.000000001}");
+        assert!(t.unwrap_err().to_string().contains(&ParseAmountError::TooPrecise.to_string()));
+        //let t: Result<T, serde_json::Error> = serde_json::from_str("{\"amt\": 21000000.000000001}");
```

Comment:
Can this be removed?

Reply:
Well, it's a test that should work, but the fact that serde forces us to do floating point makes it not work.

Reply:
If you think we'll be able to use the code in the future can you make it a `FIXME` or `TODO` please?  If not removing it seems appropriate.

Reply:
I don't think we'll be able to use it in the foreseeable future. We ought to just remove it.

Reply:
Removed.